### DISCUSSION TITLE
Auto install latex packages without rendering document

### DISCUSSION THREAD

**Discussion by ratnanil** [quarto-dev/quarto-cli#2518] [Q&A] [answered]
I would like to extend my [rocker image](https://rocker-project.org/images) to include with all the necessary latex packages preinstalled in order to complile a quarto document to [quarto-elsevier](https://github.com/quarto-journals/elsevier). Is there a way to do this automatically? 

My naïve approach would be to extract everything in between `usepackage{` and `}` from all tex files (e.g. with `perl -0777 -ne 'while(m/usepackage{(.+?)}/g){print "$1\n";}' *.tex | tr "," "\n"` and then run `tinytex::tlmgr_install()` on all of them. Is there a more elegant and simpler way?

<details>
<summary>packages for quarto-elsevier </summary>

``` 
amsmath
amssymb
lmodern
iftex
textcomp
unicode-math
upquote
parskip
xcolor
color
fancyvrb
framed
longtable
booktabs
array
calc
etoolbox
footnotehyper
footnote
graphicx
caption
float
caption
subcaption
float
selnolig
bookmark
hyperref
xurl
```

</detail>

**Comment by dragonstyle** [comment]
That seems like a reasonable approach to me. 

Quarto's implementation of TinyTeX would automatically install the packages when a journal article is rendered, but since the packages are discovered at runtime, it would only install the packages required to compile the current document in the current configuration and could theoretically miss packages that are used conditionally. So it isn't much help in this case.

In the end, since Quarto allows authors to inject packages via `include-in-header` there is no practical way to ensure that you have all the required packages that could possibly be used short of installing a full install of TeX (~4GB).

**Comment by ratnanil** [comment, marked as answer]
My manual approach did not work very well, apparently some of the results of `perl -0777 -ne 'while(m/usepackage{(.+?)}/g){print "$1\n";}' *.tex | tr "," "\n"` weren't really latex packages and other necessary packages were missing. I'm now using the following approach instead:


```sh
# create a directory with an arbitrary name and `cd` into this directory
mkdir elsevier; cd elsevier              
# run the command following command to install the desired template                                 
quarto use template quarto-journals/elsevier --no-prompt
# render the template qmd file, which has the same name as the parent folder
quarto render elsevier.qmd
# cd out of the directory and remove the created folder (we don't need it anymore)
cd ..; rm -r elsevier
```

I mashed all these commands up into one line, prepended it with `RUN` and added it to my `dockerfile`. 

```
RUN mkdir elsevier; cd elsevier; quarto use template quarto-journals/elsevier --no-prompt; quarto render elsevier.qmd; cd ..; rm -r elsevier
```

PS: Using `git clone https://github.com/quarto-journals/elsevier` might have been more elegant, I'll try this the next time

**Reply by dragonstyle** [reply]
Thanks for sharing the update! `quarto use` is more or less the same as cloning the repo (though it knows how to exclude specific files that are more about the repo than they are about the template).
